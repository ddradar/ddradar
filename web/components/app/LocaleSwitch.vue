<script setup lang="ts">
const { locale, locales } = useI18n()

const items = computed(() => [
  locales.value.map(l => ({
    label: l.name!,
    icon: `i-flag-${l.flag}-4x3`,
    click: () => (locale.value = l.code),
  })),
])
</script>

<template>
  <UDropdown :items="items">
    <UButton
      icon="i-heroicons-language-20-solid"
      color="gray"
      variant="ghost"
    />
  </UDropdown>
</template>
