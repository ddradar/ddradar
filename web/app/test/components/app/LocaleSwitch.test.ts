import { mockNuxtImport, mountSuspended } from '@nuxt/test-utils/runtime'
import type { LocaleObject } from '@nuxtjs/i18n'
import { RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'

import LocaleSwitch from '~/components/app/LocaleSwitch.vue'
import config from '~~/nuxt.config'

const { useI18nMock } = vi.hoisted(() => ({ useI18nMock: vi.fn() }))
mockNuxtImport('useI18n', () => useI18nMock)

describe('components/app/LocaleSwitch.vue', () => {
  const stubs = { RouterLink: RouterLinkStub }
  const locales = [...(config.i18n!.locales as LocaleObject[])]

  test('snapshot test', async () => {
    // Arrange
    useI18nMock.mockReturnValue({ locale: ref('ja'), locales: ref(locales) })
    const wrapper = await mountSuspended(LocaleSwitch, { global: { stubs } })
    // Act
    await wrapper.find('button').trigger('click') // Expand dropdown
    // Assert
    expect(wrapper.find('div').element).toMatchSnapshot()
  })

  test.each(locales.map((l, i) => [i, l.code]))(
    'click menu:%i changes locale to "%s"',
    async (i, code) => {
      // Arrange
      const locale = ref('bar')
      useI18nMock.mockReturnValue({ locale, locales: ref(locales) })
      const wrapper = await mountSuspended(LocaleSwitch, { global: { stubs } })
      // Act
      await wrapper.find('button').trigger('click') // Expand dropdown
      await wrapper.findAll("button[role='menuitem']")[i]?.trigger('click') // Click menu item
      // Assert
      expect(locale.value).toBe(code)
    }
  )
})
