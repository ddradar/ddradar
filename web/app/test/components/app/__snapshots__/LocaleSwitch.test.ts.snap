// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/app/LocaleSwitch.vue > snapshot test 1`] = `
<div
  class="relative inline-flex text-left rtl:text-right"
  data-headlessui-state="open"
>
  <div
    aria-controls="headlessui-menu-items-niF71ZoIms3_1"
    aria-expanded="true"
    aria-haspopup="menu"
    class="inline-flex w-full"
    data-headlessui-state="open"
    disabled="false"
    id="headlessui-menu-button-niF71ZoIms3_0"
    role="button"
  >
    
    <button
      class="focus:outline-none focus-visible:outline-0 disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 p-1.5 text-gray-700 dark:text-gray-200 hover:text-gray-900 dark:hover:text-white hover:bg-gray-50 dark:hover:bg-gray-800 focus-visible:ring-inset focus-visible:ring-2 focus-visible:ring-primary-500 dark:focus-visible:ring-primary-400 inline-flex items-center"
      type="button"
    >
      
      
      <svg
        aria-hidden="true"
        class="icon flex-shrink-0 h-5 w-5"
        data-v-60b76eb3=""
        height="1em"
        role="img"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M7.75 2.75a.75.75 0 0 0-1.5 0v1.258a33 33 0 0 0-3.599.278a.75.75 0 1 0 .198 1.487A31.6 31.6 0 0 1 8.7 5.545A19.4 19.4 0 0 1 7 9.56a19 19 0 0 1-1.002-2.05a.75.75 0 0 0-1.384.577a21 21 0 0 0 1.492 2.91a19.6 19.6 0 0 1-3.828 4.154a.75.75 0 1 0 .945 1.164A21 21 0 0 0 7 12.331q.143.198.29.391a.75.75 0 0 0 1.194-.91q-.306-.4-.59-.815a21 21 0 0 0 2.333-5.332q.465.047.924.108a.75.75 0 0 0 .198-1.487a33 33 0 0 0-3.599-.278z"
          />
          <path
            clip-rule="evenodd"
            d="M13 8a.75.75 0 0 1 .671.415l4.25 8.5a.75.75 0 1 1-1.342.67L15.787 16h-5.573l-.793 1.585a.75.75 0 1 1-1.342-.67l4.25-8.5A.75.75 0 0 1 13 8m2.037 6.5L13 10.427L10.964 14.5z"
            fill-rule="evenodd"
          />
        </g>
      </svg>
      
      
      <!--v-if-->
      
      
      <!--v-if-->
      
      
    </button>
    
  </div>
  <div
    class="z-20 group w-48"
    data-popper-placement="bottom-end"
    style="position: fixed; left: auto; top: 0px; margin: 0px; right: 0px; bottom: auto; transform: translate(0px, 8px);"
  >
    <transition-stub
      appear="true"
      css="true"
      enteractiveclass="transition duration-100 ease-out"
      enterfromclass="transform scale-95 opacity-0"
      entertoclass="transform scale-100 opacity-100"
      leaveactiveclass="transition duration-75 ease-in"
      leavefromclass="transform scale-100 opacity-100"
      leavetoclass="transform scale-95 opacity-0"
      persisted="false"
    >
      <div>
        <!--v-if-->
        <div
          aria-labelledby="headlessui-menu-button-niF71ZoIms3_0"
          class="relative focus:outline-none overflow-y-auto scroll-py-1 divide-y divide-gray-200 dark:divide-gray-700 ring-1 ring-gray-200 dark:ring-gray-700 rounded-md shadow-lg bg-white dark:bg-gray-800"
          data-headlessui-state="open"
          id="headlessui-menu-items-niF71ZoIms3_1"
          role="menu"
          tabindex="0"
        >
          
          <div
            class="p-1"
            role="none"
          >
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-niF71ZoIms3_2"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                English
              </span>
              <!--v-if-->
              
            </button>
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-niF71ZoIms3_3"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                日本語
              </span>
              <!--v-if-->
              
            </button>
            
            
          </div>
          
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
